### PR TITLE
Support NullEQ join keys

### LINKED ISSUE #10787: Support null-eq join
## Enhancement

Support null-eq join in TiFlash.

This includes:
- plumbing null-eq join metadata from DAG/planner to join execution
- handling nullable null-eq join keys correctly in hash join
- separating row filtering from key-null handling in outer/full join paths
- disabling incompatible runtime filter cases for nullable null-eq keys
- adding targeted test coverage for null-eq join execution paths

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #10787

Problem Summary:

TiFlash hash join currently filters nullable join keys before probing or building the hash map, which makes `NULL <=> NULL` unable to match. This PR adds join-key-level NullEQ semantics while preserving ordinary equality semantics for other keys.

### What is changed and how it works?

```commit-message
Support NullEQ join keys in TiFlash hash join.
```

- Parse and propagate `is_null_eq` from the join request to the execution layer.
- Preserve nullable NullEQ keys during build and probe, while continuing to filter NULLs from ordinary `=` keys.
- Align mixed nullable and non-nullable key schemas.
- Update outer join, scan-after-probe, spill, and fine-grained shuffle paths.
- Disable incompatible runtime-filter paths and add coverage for join semantics and key-map selection.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

## Summary by CodeRabbit

* **New Features**
  * Added per-key null-safe equality (`<=>`) support for hash joins.
  * NULL values now match correctly for null-safe keys while ordinary equality behavior remains unchanged.
  * Supports mixed key types, outer and semi joins, spill processing, and fine-grained shuffle.

* **Bug Fixes**
  * Improved NULL filtering and join-key handling.
  * Prevented incompatible runtime filters for nullable null-safe keys.

* **Documentation**
  * Added documentation covering null-safe join behavior, supported scenarios, and limitations.

### CODE REVIEW

File: dbms/src/Interpreters/JoinHashMap.h [hunk 1]
```diff
@@ -171,6 +173,8 @@ struct ConcurrentMapsTemplate
     using key_fixed_stringType = ConcurrentHashMapWithSavedHash<StringRef, Mapped>;
     using keys128Type = ConcurrentHashMap<UInt128, Mapped, HashCRC32<UInt128>>;
     using keys256Type = ConcurrentHashMap<UInt256, Mapped, HashCRC32<UInt256>>;
+    using nullable_keys128Type = ConcurrentHashMap<UInt128, Mapped, HashCRC32<UInt128>>;
```
Comment (gengliqi) [inline dbms/src/Interpreters/JoinHashMap.h:176]: It seems wasteful that `Nullable(UInt32)` will use `UInt128` as join key.

**Review by gengliqi — COMMENTED** [review]
Rest LGTM

File: dbms/src/Interpreters/JoinHashMap.h (hunk 1 above)
Comment (windtalker) [reply dbms/src/Interpreters/JoinHashMap.h:176]: Actually, this logic was copied over from HashAgg, so HashAgg has a similar issue as well. I'll leave a TODO for now and address it later

File: dbms/src/Interpreters/Join.cpp [hunk 2]
```diff
@@ -399,8 +456,18 @@ void Join::initBuild(const Block & sample_block, size_t build_concurrency_)
...
+    join_map_method = chooseJoinMapMethod(
+        getKeyColumns(key_names_right, sample_block, is_null_eq),
+        key_sizes,
+        collators,
+        is_null_eq);
+    if (hasNullEqKey(is_null_eq))
+    {
+        if (join_map_method == JoinMapMethod::serialized)
+            LOG_DEBUG(log, "Use serialized join map method because nullable NullEQ keys do not fit packed fixed keys");
+    }
```
Comment (gengliqi) [inline dbms/src/Interpreters/Join.cpp:470]: Seems not necessary？

**Review by gengliqi — APPROVED** [review]
LGTM

**Review by solotzg — APPROVED** [review]
